velox/exec/fuzzer/WindowFuzzer.h

```diff
@@ -79,14 +79,88 @@ class WindowFuzzer : public AggregationFuzzerBase {
   void go(const std::string& planPath);
 
  private:
+  struct FrameMetadata {
+    core::WindowNode::WindowType windowType =
+        core::WindowNode::WindowType::kRange;
+    core::WindowNode::BoundType startBoundType =
+        core::WindowNode::BoundType::kUnboundedPreceding;
+    core::WindowNode::BoundType endBoundType =
+        core::WindowNode::BoundType::kCurrentRow;
+    std::string startBoundString = "";
+    std::string endBoundString = "";
+  };
+
   void addWindowFunctionSignatures(const WindowFunctionMap& signatureMap);
 
-  // Return a randomly generated frame clause string together with a boolean
-  // flag indicating whether it is a ROWS frame.
-  std::string generateFrameClause(
+  // Generates the frame bound for k-ROWS frames, decides randomly if the frame
+  // bound is a constant or a column. Returns the constant integer value when
+  // K is a constant, otherwise returns the column name used as frame bound.
+  const std::string& generateKRowsFrameBound(
+      std::vector<std::string>& argNames,
+      std::vector<TypePtr>& argTypes,
+      const std::string& columnName);
+
+  // Randomly generates the frame metadata, which includes the window type, the
+  // frame start and end bound type, and the constant/column offset for k-ROWS
+  // frames. For k-RANGE frames, the constant/column offset generation is
+  // deferred till the ORDER-BY key is generated. The FrameMetadata returned by
+  // this function is used in sorting key(s) generation, to ensure a single
+  // ORDER-BY key exists for k-RANGE frames.
+  FrameMetadata generateFrameClause(
       std::vector<std::string>& argNames,
       std::vector<TypePtr>& argTypes,
-      bool& isRowsFrame);
+      const std::vector<std::string>& kBoundColumnNames);
+
+  // Adds offset column(s) to input data for k-RANGE frames using the helper
+  // function addKRangeOffsetColumnToInputImpl.
+  template <TypeKind TKind>
+  void addKRangeOffsetColumnToInput(
+      std::vector<RowVectorPtr>& input,
+      FrameMetadata& frameMetadata,
+      SortingKeyAndOrder& orderByKey,
+      const std::vector<std::string>& columnNames,
+      const std::vector<std::string>& offsetColumnNames);
+
+  // Adds offset column to input data for k-range frames. Randomly decides if
+  // the frame bound will be a constant or a column.
+  // 1. If the frame bound is a constant, the constant K value is randomly
+  // generated to be of the same type as the ORDER-BY column. The offset column
+  // value for a given row is then generated by helper function genOffsetAtIdx.
+  // The constant K value is returned as a string.
+  // 2. If the frame bound is a column, VectorFuzzer is used to generate a
+  // column of the same type as the ORDER-BY key. Returns offset column name.
```

Comment:
Looks like we only generate constant frame bound columns right now. Could we clarify this in the comment so that we remember to extend it later?